### ISSUE TITLE
Where are episodes converted to batches?

### ISSUE THREAD

**Issue by zhenduow** [facebookresearch/ParlAI#3418]
Hi,

I am trying to understand the poly-encoder training code and potentially hacking it with some structure changes. I need to add more field in data and be able to parse them into batches for the new structure. 

I see that ParlAIDialogTeacher reads the data and store them in `self.episode`, but I cannot find the function that converts `ParlAIDialogTeacher.episode` to batches observations which contain the texts and tokenizations. I suppose it happens somewhere in `world.create_task_world()`, but I can't find it. Can you point me to the function? 

Thank you very much!

Zhenduo

**Comment by stephenroller** [comment]
Ah our setup is quite complex due to the Agents "talking" back and forth.

If you want to add a new field to be readable by the polyencoder, there are a couple of steps:
- add the field into the training data. Since you're using a parlaidialogteacher, I think you've probably figured out how to do that. (just append \t field_name:field value to each line of the file)
- next, you want to override `batchify`. As an example, here's adding the "rewards" field in one agent:

https://github.com/facebookresearch/ParlAI/blob/2b4ced6e812b8aacfa4e956fa86d42dbccdf2fcf/projects/dialogue_unlikelihood/agents.py#L69-L75

This on its own will make sure the field is in the Batch object passed to train_step and eval_step, but it's still up to you to figure out what to do with that.

If you want to actually hack _batching_ (that is, grouping multiple samples together), that's all in BatchWorld. You can reading our [batching tutorial](https://parl.ai/docs/tutorial_worlds.html) to get a high level overview of how that works. In short, it's just grouping random examples together, and it would be unusual to want to modify this.

**Comment by zhenduow** [comment]
> Ah our setup is quite complex due to the Agents "talking" back and forth.
> 
> If you want to add a new field to be readable by the polyencoder, there are a couple of steps:
> 
> * add the field into the training data. Since you're using a parlaidialogteacher, I think you've probably figured out how to do that. (just append \t field_name:field value to each line of the file)
> * next, you want to override `batchify`. As an example, here's adding the "rewards" field in one agent:
> 
> https://github.com/facebookresearch/ParlAI/blob/2b4ced6e812b8aacfa4e956fa86d42dbccdf2fcf/projects/dialogue_unlikelihood/agents.py#L69-L75
> 
> This on its own will make sure the field is in the Batch object passed to train_step and eval_step, but it's still up to you to figure out what to do with that.
> 
> If you want to actually hack _batching_ (that is, grouping multiple samples together), that's all in BatchWorld. You can reading our [batching tutorial](https://parl.ai/docs/tutorial_worlds.html) to get a high level overview of how that works. In short, it's just grouping random examples together, and it would be unusual to want to modify this.

Thank you, Stephen! 

Yes, I have figured out how to override `batchify` and added fields to be read in the teacher. The output of teacher has the new fields I add. But, when I try to print the input of batchify `obs_batch`, I find that my new fields are removed and also all fields that are texts are tokenized. Where should I change so that my new fields are not removed and can be tokenized?

**Comment by stephenroller** [comment]
Tokenization happens inside observe. However, the input to batchify should be unmodified messages, so if you've added extra fields, you should be able to see them

One good way to sanity check is to use `parlai display_data -t yourtask ... --verbose`, which will show you every field that the teacher is emitting.

I'm afraid i'll need more specifics before I can help you further than this.

**Comment by zhenduow** [comment]
> Tokenization happens inside observe. However, the input to batchify should be unmodified messages, so if you've added extra fields, you should be able to see them
> 
> One good way to sanity check is to use `parlai display_data -t yourtask ... --verbose`, which will show you every field that the teacher is emitting.
> 
> I'm afraid i'll need more specifics before I can help you further than this.

Yes, you are right. I checked the output of `display_data` and found the reason. Now the input to batchify has all my fields. Thank you! 

Also, where is the `observe` that also tokenizes texts implemented? In world.py `ExecutableWorld.observe`, there isn't tokenization. I guess that is not the real `observe` function...

Thank you!

**Comment by stephenroller** [comment]
Here, in torchagent.observe:

https://github.com/facebookresearch/ParlAI/blob/2b4ced6e812b8aacfa4e956fa86d42dbccdf2fcf/parlai/core/torch_agent.py#L1689-L1708

Note that `update_history` is where we tokenize the context, but self.vectorize is where we tokenize the label and any other special fields. `update_history` is done separately so that we can cache the previous tokenization from turn-to-turn for speedups.

**Comment by zhenduow** [comment]
Thank you! Separating them seems reasonable. I probably only need to add some lines in `vectorize`.

Thank you for explaining all the secrets to me!

**Comment by zhenduow** [comment]
> Here, in torchagent.observe:
> 
> https://github.com/facebookresearch/ParlAI/blob/2b4ced6e812b8aacfa4e956fa86d42dbccdf2fcf/parlai/core/torch_agent.py#L1689-L1708
> 
> Note that `update_history` is where we tokenize the context, but self.vectorize is where we tokenize the label and any other special fields. `update_history` is done separately so that we can cache the previous tokenization from turn-to-turn for speedups.

Thank you! Separating them seems reasonable. I probably only need to add some lines in vectorize.

Thank you for explaining all the secrets to me!

**Comment by stephenroller** [comment]
Closing the issue but you can reopen if you have further questions.

**Comment by zhenduow** [comment]
Hi Stephen, 

I change the structure of polyencoder and it runs now. But during eval: valid, it loses 'accuracy' metric in report. It still has some other metrics including (exs  gpu_mem  loss    lr   mrr  rank  total_train_updates  tpb). My question is where are these computed so that I can check why 'accuracy' is lost.

**Comment by zhenduow** [comment]
> Closing the issue but you can reopen if you have further questions.

Hi Stephen,

I change the structure of polyencoder and it runs now. But during eval: valid, it loses 'accuracy' metric in report. It still has some other metrics including (exs gpu_mem loss lr mrr rank total_train_updates tpb). My question is where are these computed so that I can check why 'accuracy' is lost.

**Comment by stephenroller** [comment]
Does it have hits@1? This suggests to me that `eval_step` is returning an empty `Output` object

**Comment by zhenduow** [comment]
No, it does not have hits@1. I make my own version of `eval_step` and redefined some of the metrics like mrr, rankn and loss, and my returning `Output` is also redefined as super of `Output`. But I didn't see accuracy or hits@1 in the original `eval_step`. So, I guess they are computed elsewhere. Where should I compute them then?

**Comment by zhenduow** [comment]
> Does it have hits@1? This suggests to me that `eval_step` is returning an empty `Output` object

No, it does not have hits@1. I make my own version of eval_step and redefined some of the metrics like mrr, rankn and loss, and my returning Output is also redefined as super of Output. But I didn't see accuracy or hits@1 in the original eval_step. So, I guess they are computed elsewhere. Where should I compute them then? 

Thanks!

**Comment by stephenroller** [comment]
Accuracy is a teacher-side metric. Your model needs to return a "text" field in Output for it to be used.

**Comment by zhenduow** [comment]
> Accuracy is a teacher-side metric. Your model needs to return a "text" field in Output for it to be used.

I changed the returned field name to text_1 and text_2. So that’s why teacher doesn’t compute accuracy. Which teacher function I change to compute accuracy manually?

**Comment by stephenroller** [comment]
Strictly speaking here:

https://github.com/facebookresearch/ParlAI/blob/72e8e6157d1905f5e16b762279688479850a0843/parlai/core/metrics.py#L789-L794

Another option you could consider is using [Teacher Metrics](https://parl.ai/docs/tutorial_metrics.html#teacher-metrics) if you're using your own teacher.
